controllers/controller_shared.go

```diff
@@ -58,6 +58,85 @@ func GetMatchingInstances(ctx context.Context, k8sClient client.Client, labelSel
 	return selectedList, err
 }
 
+// Only matching instances in the scope of the resource are returned
+// Resources with allowCrossNamespaceImport expands the scope to the entire cluster
+// Intended to be used in reconciler functions
+func GetScopedMatchingInstances(log logr.Logger, ctx context.Context, k8sClient client.Client, cr v1beta1.CommonResource) ([]v1beta1.Grafana, error) {
+	instanceSelector := cr.MatchLabels()
+
+	if instanceSelector.MatchLabels == nil {
+		return []v1beta1.Grafana{}, nil
+	}
```

Comment:
An empty label selector should match _all_ instances, not none. At least this is the behavior right now, and I think there's value in keeping it that way

Reply:
From what I can tell that's not the current behaviour, the existing implementation of [`GetMatchingInstances(...)`](https://github.com/grafana/grafana-operator/pull/1770/files#diff-4f9e711de4a512202068ca40cbb5a0f8ce367474278ececd6567c8ba60473f01R37-R59) exits early.
I left that function alone to not interfere with any other reconciliation loop.
Additionally, instanceSelector is not an `optional/nullable` field:
![image](https://github.com/user-attachments/assets/ef3e93b9-c055-4289-872d-4286b07e717c)

Reply:
Yeah, don't merge this yet, I need to fully understand instanceSelector behaviour before this is ready.

Reply:
@theSuess Updated the branch a couple of days ago, addressing the regressions and the tests.
After letting it sit for a few days and going back over it, I think it's ready.

I'm not a huge fan of `instanceSelector` as it's implemented right now and I will likely open a PR Changing the behaviour later on.
specifically making the field `optional` and `nullable` to align the CRD with the implemented behaviour.